src/document/doc-utils.js

```diff
@@ -180,42 +180,99 @@ function removeLines(doc) {
-  for (let i = doc.parts.length; i > 0 && !lastPart; i--) {
-    lastPart = parts[i - 1];
+const isHardline = (doc, nextDoc) =>
+  doc &&
+  doc.type === "line" &&
+  doc.hard &&
+  nextDoc &&
+  nextDoc.type === "break-parent";
+function stripDocTrailingHardlineFromDoc(doc) {
+  if (!doc) {
+    return doc;
   }
-  if (lastPart.type === "group") {
-    parts = lastPart.contents.parts;
+
+  switch (doc.type) {
+    case "concat":
+    case "fill": {
+      const { parts } = doc;
+
+      while (parts.length > 1 && isHardline(...parts.slice(-2))) {
+        parts.pop();
+        parts.pop();
+      }
+
+      if (parts.length > 1) {
+        const lastPart = stripDocTrailingHardlineFromDoc(
+          parts[parts.length - 1]
+        );
+        parts[parts.length - 1] = lastPart;
+      }
+      break;
+    }
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      doc.contents = stripDocTrailingHardlineFromDoc(doc.contents);
+      break;
+    case "if-break":
+      doc.breakContents = stripDocTrailingHardlineFromDoc(doc.breakContents);
+      doc.flatContents = stripDocTrailingHardlineFromDoc(doc.flatContents);
+      break;
   }
-  return parts;
+
+  return doc;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
+function stripTrailingHardline(doc) {
   // HACK remove ending hardline, original PR: #1984
-  if (doc.type === "concat" && doc.parts.length !== 0) {
-    const parts = withInnerParts ? getInnerParts(doc) : doc.parts;
-    const lastPart = parts[parts.length - 1];
-    if (lastPart.type === "concat") {
-      if (
-        lastPart.parts.length === 2 &&
-        lastPart.parts[0].hard &&
-        lastPart.parts[1].type === "break-parent"
-      ) {
-        return { type: "concat", parts: parts.slice(0, -1) };
-      }
+  return stripDocTrailingHardlineFromDoc(cleanDoc(doc));
+}
 
-      return {
-        type: "concat",
-        parts: doc.parts.slice(0, -1).concat(stripTrailingHardline(lastPart)),
-      };
+const isConcat = (doc) => doc && doc.type === "concat";
+function cleanDocFn(doc) {
+  if (!isConcat(doc)) {
+    return doc;
+  }
+
+  const { parts } = doc;
+  for (let index = parts.length - 1; index >= 0; index--) {
+    let currentDoc = cleanDocFn(parts[index]);
+    // Next doc already normalized
+    const nextDoc = parts[index + 1];
+
+    // Concat continuous string
+    if (typeof currentDoc === "string" && typeof nextDoc === "string") {
+      parts.splice(index + 1, 1);
+      currentDoc += nextDoc;
+    }
+    // Concat continuous `concat`
+    else if (isConcat(currentDoc) && isConcat(nextDoc)) {
+      parts.splice(index + 1, 1);
+      currentDoc.parts.push(...nextDoc.parts);
+    }
+
+    // If empty string, remove it
+    if (currentDoc === "") {
+      parts.splice(index, 1);
+    } else {
+      parts[index] = currentDoc;
     }
   }
 
+  if (parts.length === 0) {
+    return "";
+  }
+
+  if (parts.length === 1) {
+    return cleanDocFn(parts[0]);
```

Comment:
Why call `cleanDocFn` here again? Aren't all the parts already processed at this point?

Reply:
I guess I because I didn't turn `{type: "concat", parts: []}` into `""` at first, I'll check.

Reply:
Overall, it's probably just me, but it's hard to follow what's going on in this function as compared with `normalizeDoc`, which looks much simpler. Can we probably just use `normalizeDoc` with an extra parameter to protect `fill`?

Reply:
`normalizeDoc` didn't remove empty `concat`, so we can't locate hardLine.

Reply:
If I correctly understand what you mean, originally it did that:

https://github.com/prettier/prettier/pull/8496/files#diff-751b7c98ecb8b8f58f17f22b778c467270f4d0a4fd6b72a262c71a8a243888b6L860-L862

But that was somehow lost in refactoring. Try adding it back?

Reply:
Example

```json
{
	type:"concat",
	parts: [
		{
			type: 'concat'
			[
				{type: "concat", parts: [...HARD_LINE]}
			],
		}
		// A empty string
		{
			type: 'concat',
			parts: ['']
		}
	]
}
```

Reply:
Looks like exactly what those removed lines I linked to were for.

Reply:
> But that was somehow lost in refactoring. Try adding it back?

I  prefer use the new one. It do much more normalize than the old one, we can extend it to let it work for `fill` if we want replace old one later.

More clean, more easy for finding these new lines.